File: native/ios/FallbackWidgets/Snapshot.swift
//
//  Snapshot.swift
//  FallbackWidgets
//
//  M7. Decodes the JSON snapshot `src/services/widgets/index.ts` writes into the
//  `group.com.fallback.app` App Group container (see `Paths.appleSharedContainers` in
//  that file). This is the WIDGET'S ONLY DATA SOURCE — it never opens the app's SQLite
//  database (MODULES.md M7's non-negotiable), and every field here is already fully
//  resolved (display strings, hex colours) by the JS side, so this file contains no
//  business logic of its own — just decoding + rendering.
//
import Foundation

struct WidgetSnapshot: Codable {
    struct Theme: Codable {
        let scheme: String   // "light" | "dark"
        let accent: String   // AccentKey
        let accentHex: String
    }

    struct SignalColors: Codable {
        let ideal: String
        let fallback: String
        let off: String
        // Deliberately no `missed` field — missed has no fill colour anywhere in the app
        // (Rule 4 / ARCHITECTURE §5), and the JS-side snapshot never sends one.
    }

    struct Today: Codable {
        let totalDue: Int
        let doneCount: Int
        let percent: Int?
    }

    struct TaskSummary: Codable {
        let id: String
        let name: String
        let timeOfDay: String?
        let chipState: String?  // "todo" | "done" | "fallback" | "skip" | null
    }

    struct WidgetConfigEntry: Codable {
        let size: String        // "small-today" | "small-one-task" | "medium-up-next"
        let mode: String        // "fixed-task" | "smart-next-due"
        let fixedTaskId: String?
    }

    let version: Int
    let generatedAt: String
    let theme: Theme
    let signalColors: SignalColors
    let today: Today
    let tasks: [TaskSummary]
    let widgetConfigs: [WidgetConfigEntry]

    /// The shared-container file the JS side writes. Matches `SNAPSHOT_FILENAME` /
    /// `APP_GROUP_ID` in `src/services/widgets/index.ts` exactly — if either changes there,
    /// this constant must change too.
    static let appGroupId = "group.com.fallback.app"
    static let filename = "fallback-widget-snapshot.json"

    static func load() -> WidgetSnapshot? {
        guard
            let containerURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupId),
            let data = try? Data(contentsOf: containerURL.appendingPathComponent(filename))
        else {
            return nil
        }
        return try? JSONDecoder().decode(WidgetSnapshot.self, from: data)
    }

    /// Which task a given widget config should surface: the fixed pick if set and still due
    /// today, else the first due task ("smart — next due" — S46's own definition, "whichever
    /// task is due soonest," approximated here as "first in the resolved due-today list,"
    /// since ordering by literal due time is display-only formatting the JS side already
    /// applies to `tasks` before writing the snapshot).
    func task(for config: WidgetConfigEntry) -> TaskSummary? {
        if config.mode == "fixed-task", let fixedId = config.fixedTaskId {
            return tasks.first(where: { $0.id == fixedId }) ?? tasks.first
        }
        return tasks.first
    }

    func config(for size: String) -> WidgetConfigEntry {
        widgetConfigs.first(where: { $0.size == size })
            ?? WidgetConfigEntry(size: size, mode: "smart-next-due", fixedTaskId: nil)
    }
}

File: src/services/widgets/snapshot.ts
/**
 * M7. F21 — pure computation of the shared-container JSON snapshot the native widget
 * targets read. No I/O here: `index.ts` does the actual file write; this module only shapes
 * the data, so it is unit-testable without a filesystem.
 *
 * The native targets "never open the SQLite database directly" (MODULES.md M7) — this
 * snapshot is their ONLY input, which is why it carries pre-resolved display strings and
 * pre-resolved colours (not raw task/log rows) rather than pushing any business logic onto
 * the Swift/Kotlin side.
 */
import type { ColorScheme, Palette } from '@/theme';
import type { AccentKey, ChipState, LocalDate, OccurrenceOutcome, TaskWithSteps, WidgetConfig } from '@/types';

export interface WidgetSignalColors {
  readonly ideal: string;
  readonly fallback: string;
  readonly off: string;
}

export interface WidgetTaskSummary {
  readonly id: string;
  readonly name: string;
  readonly timeOfDay: string | null;
  /** `null` — nothing due yet today for this task; the native side renders the `todo` visual. */
  readonly chipState: ChipState | null;
}

export interface WidgetSnapshot {
  readonly version: 1;
  readonly generatedAt: string;
  readonly theme: { readonly scheme: ColorScheme; readonly accent: AccentKey; readonly accentHex: string };
  /** Rule 4 / §5 — signal colours are accent- and (for display purposes) theme-resolved already; missed has none. */
  readonly signalColors: WidgetSignalColors;
  readonly today: { readonly totalDue: number; readonly doneCount: number; readonly percent: number | null };
  /** Every DUE task today, chip-resolved, for the "smart — next due" / "fixed task" widget configs to select from. */
  readonly tasks: readonly WidgetTaskSummary[];
  readonly widgetConfigs: readonly WidgetConfig[];
}

export interface ResolvedTaskChip {
  readonly taskId: string;
  readonly chipState: ChipState | null;
  /**
   * Review pass 1, blocking item 3: the caller (`index.ts`) resolves this through
   * `resolveOccurrence`/`designateCarrier` (movedInLog-aware), not raw `isDue` — a task
   * snoozed AWAY from today resolves `'not-due'` here even though cadence still says today is
   * due; a task snoozed INTO today resolves a real due outcome even though cadence alone says
   * it is not. `dueTasks` below filters on THIS field, never on `isDue` directly.
   */
  readonly outcome: OccurrenceOutcome;
}

/**
 * `chips` — the day's resolved chip per due task, supplied by the caller (`index.ts`, which
 * has repo access). Kept as a plain parameter so this function stays a pure data shape, not a
 * second read-path implementation.
 */
export function buildWidgetSnapshot(input: {
  readonly tasks: readonly TaskWithSteps[];
  readonly chips: readonly ResolvedTaskChip[];
  readonly date: LocalDate;
  readonly scheme: ColorScheme;
  readonly accent: AccentKey;
  readonly accentHex: string;
  readonly palette: Palette;
  readonly widgetConfigs: readonly WidgetConfig[];
  readonly generatedAt: string;
}): WidgetSnapshot {
  const { tasks, chips, scheme, accent, accentHex, palette, widgetConfigs, generatedAt } = input;
  const chipByTask = new Map(chips.map((c) => [c.taskId, c]));

  const dueTasks = tasks.filter((t) => {
    if (t.deletedAt) return false;
    const resolved = chipByTask.get(t.id);
    return resolved !== undefined && resolved.outcome !== 'not-due';
  });
  const summaries: WidgetTaskSummary[] = dueTasks.map((t) => ({
    id: t.id,
    name: t.name,
    timeOfDay: t.timeOfDay,
    chipState: chipByTask.get(t.id)?.chipState ?? null,
  }));

  const doneCount = summaries.filter((s) => s.chipState === 'done' || s.chipState === 'fallback').length;
  const totalDue = summaries.length;

  return {
    version: 1,
    generatedAt,
    theme: { scheme, accent, accentHex },
    signalColors: { ideal: palette.ideal, fallback: palette.fallback, off: palette.off },
    today: { totalDue, doneCount, percent: totalDue > 0 ? Math.round((doneCount / totalDue) * 100) : null },
    tasks: summaries,
    widgetConfigs,
  };
}
